File: lecture-notes/app-dev-pathway/week-02-design-patterns/code-examples/factory.py
import random


class Card:
    def __init__(self, suit, value):
        self.suit = suit
        self.value = value

    def __str__(self):
        names = {11: "Jack", 12: "Queen", 13: "King", 14: "Ace"}
        name = names.get(self.value, str(self.value))
        return f"{name} of {self.suit}"


class IPlayer:
    def get_name(self):
        raise NotImplementedError

    def play_card(self):
        raise NotImplementedError

    def receive_card(self, card):
        raise NotImplementedError

    def has_cards(self):
        raise NotImplementedError

    def deck_size(self):
        raise NotImplementedError


class HumanPlayer(IPlayer):
    def __init__(self, name):
        self.name = name
        self.hand = []

    def get_name(self):
        return self.name

    def has_cards(self):
        return len(self.hand) > 0

    def deck_size(self):
        return len(self.hand)

    def play_card(self):
        input(f"{self.name}, press ENTER to draw your card...")
        card = self.hand.pop()
        print(f"{self.name} draws: {card}")
        return card

    def receive_card(self, card):
        self.hand.insert(0, card)


class ComputerPlayer(IPlayer):
    def __init__(self, name):
        self.name = name
        self.hand = []

    def get_name(self):
        return self.name

    def has_cards(self):
        return len(self.hand) > 0

    def deck_size(self):
        return len(self.hand)

    def play_card(self):
        card = self.hand.pop()
        print(f"{self.name} (CPU) draws: {card}")
        return card

    def receive_card(self, card):
        self.hand.insert(0, card)


class IPlayerFactory:
    def create_player(self, name):
        raise NotImplementedError


class HumanPlayerFactory(IPlayerFactory):
    def create_player(self, name):
        return HumanPlayer(name)


class ComputerPlayerFactory(IPlayerFactory):
    def create_player(self, name):
        return ComputerPlayer(name)


class FactoryWarGame:
    def start_game(self, factory1, factory2):
        p1 = factory1.create_player("Alice")
        p2 = factory2.create_player("Bob (CPU)")

        suits = ["Spades", "Hearts", "Clubs", "Diamonds"]
        deck = [Card(suit, value) for suit in suits for value in range(2, 15)]
        random.shuffle(deck)

        for card in deck[:26]:
            p1.receive_card(card)
        for card in deck[26:]:
            p2.receive_card(card)

        round_num = 1
        while p1.has_cards() and p2.has_cards() and round_num <= 10:
            print(f"\n── Round {round_num} ─────────────────────────")
            c1 = p1.play_card()
            c2 = p2.play_card()

            if c1.value > c2.value:
                p1.receive_card(c1)
                p1.receive_card(c2)
                print(f"{p1.get_name()} wins!")
            elif c2.value > c1.value:
                p2.receive_card(c1)
                p2.receive_card(c2)
                print(f"{p2.get_name()} wins!")
            else:
                p1.receive_card(c1)
                p2.receive_card(c2)
                print("Tie!")

            print(f"Cards: {p1.get_name()}={p1.deck_size()} {p2.get_name()}={p2.deck_size()}")
            round_num += 1


cpu_factory1 = ComputerPlayerFactory()
cpu_factory2 = ComputerPlayerFactory()

game = FactoryWarGame()
game.start_game(cpu_factory1, cpu_factory2)

File: lecture-notes/games-dev-pathway/week-02-design-patterns/code-examples/factory.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>
#include <memory>
#include <string>
#include <format>

struct Card
{
    std::string suit;
    int value;
    std::string toString() const
    {
        std::string name;
        if (value == 11)
            name = "Jack";
        else if (value == 12)
            name = "Queen";
        else if (value == 13)
            name = "King";
        else if (value == 14)
            name = "Ace";
        else
            name = std::to_string(value);
        return name + " of " + suit;
    }
};

class IPlayer
{
public:
    virtual std::string getName() const = 0;
    virtual Card playCard() = 0;
    virtual void receiveCard(const Card &c) = 0;
    virtual bool hasCards() const = 0;
    virtual int deckSize() const = 0;
    virtual ~IPlayer() = default;

protected:
    std::vector<Card> hand;
};

class HumanPlayer : public IPlayer
{
public:
    explicit HumanPlayer(std::string n) : name(std::move(n)) {}

    std::string getName() const override { return name; }
    bool hasCards() const override { return !hand.empty(); }
    int deckSize() const override { return static_cast<int>(hand.size()); }

    Card playCard() override
    {
        std::cout << name << ", press ENTER to draw your card...";
        std::cin.ignore();
        Card top = hand.back();
        hand.pop_back();
        std::cout << name << " draws: " << top.toString() << "\n";
        return top;
    }

    void receiveCard(const Card &c) override { hand.insert(hand.begin(), c); }

private:
    std::string name;
};

class ComputerPlayer : public IPlayer
{
public:
    explicit ComputerPlayer(std::string n) : name(std::move(n)) {}

    std::string getName() const override { return name; }
    bool hasCards() const override { return !hand.empty(); }
    int deckSize() const override { return static_cast<int>(hand.size()); }

    Card playCard() override
    {
        Card top = hand.back();
        hand.pop_back();
        std::cout << name << " (CPU) draws: " << top.toString() << "\n";
        return top;
    }

    void receiveCard(const Card &c) override { hand.insert(hand.begin(), c); }

private:
    std::string name;
};

class IPlayerFactory
{
public:
    virtual std::unique_ptr<IPlayer> createPlayer(const std::string &name) = 0;
    virtual ~IPlayerFactory() = default;
};

class HumanPlayerFactory : public IPlayerFactory
{
public:
    std::unique_ptr<IPlayer> createPlayer(const std::string &name) override
    {
        return std::make_unique<HumanPlayer>(name);
    }
};

class ComputerPlayerFactory : public IPlayerFactory
{
public:
    std::unique_ptr<IPlayer> createPlayer(const std::string &name) override
    {
        return std::make_unique<ComputerPlayer>(name);
    }
};

class FactoryWarGame
{
public:
    void startGame(IPlayerFactory &factory1, IPlayerFactory &factory2)
    {
        auto p1 = factory1.createPlayer("Alice");
        auto p2 = factory2.createPlayer("Bob (CPU)");

        std::vector<Card> deck;
        for (const auto &suit : {"Spades", "Hearts", "Clubs", "Diamonds"})
            for (int v = 2; v <= 14; ++v)
                deck.push_back({suit, v});
        std::mt19937 rng(std::random_device{}());
        std::shuffle(deck.begin(), deck.end(), rng);

        for (int i = 0; i < 26; ++i)
            p1->receiveCard(deck[i]);
        for (int i = 26; i < 52; ++i)
            p2->receiveCard(deck[i]);

        int round = 1;
        while (p1->hasCards() && p2->hasCards() && round <= 10)
        {
            std::cout << "\n── Round " << round << " ─────────────────────────\n";
            Card c1 = p1->playCard();
            Card c2 = p2->playCard();

            if (c1.value > c2.value)
            {
                p1->receiveCard(c1);
                p1->receiveCard(c2);
                std::cout << p1->getName() << " wins\n";
            }
            else if (c2.value > c1.value)
            {
                p2->receiveCard(c1);
                p2->receiveCard(c2);
                std::cout << p2->getName() << " wins\n";
            }
            else
            {
                p1->receiveCard(c1);
                p2->receiveCard(c2);
                std::cout << "Tie\n";
            }

            std::cout << std::format("Cards: {}={} {}={}\n",
                                     p1->getName(), p1->deckSize(), p2->getName(), p2->deckSize());
            ++round;
        }
    }
};

int main()
{

    ComputerPlayerFactory cpuFactory;
    ComputerPlayerFactory cpuFactory2;

    FactoryWarGame game;
    game.startGame(cpuFactory, cpuFactory2);

    return 0;
}